File: nostos-core/src/main/java/sncf/oui/nostos/core/coordinates/Coordinates.java
package sncf.oui.nostos.core.coordinates;

import org.immutables.value.Value.Immutable;
import org.immutables.value.Value.Parameter;

@Immutable
abstract class Coordinates {

    public static Coordinates of(final double latitude, final double longitude) {
        return ImmutableCoordinates.of(Latitude.of(latitude), Longitude.of(longitude));
    }

    @Parameter
    abstract Latitude latitude();

    @Parameter
    abstract Longitude longitude();

    public double getLatitude() {
        return latitude().degrees();
    }

    public double getLongitude() {
        return longitude().degrees();
    }
}

File: nostos-core/src/main/kotlin/io/fries/nostos/core/coordinates/Coordinates.kt
package io.fries.nostos.core.coordinates

class Coordinates private constructor(latitude: Latitude, longitude: Longitude) {

    val latitude = latitude.degrees
    val longitude = longitude.degrees

    companion object {
        fun of(latitude: Double, longitude: Double): Coordinates {
            return Coordinates(Latitude(latitude), Longitude(longitude))
        }
    }
}